stellargraph/mapper/graphwave_generator.py

```diff
@@ -0,0 +1,179 @@
+    """
+    Implementation of the GraphWave structural embedding algorithm from the paper:
+        "Learning Structural Node Embeddings via Diffusion Wavelets" (https://arxiv.org/pdf/1710.10321.pdf)
+
+    This class is minimally initialized with a StellarGraph object. Calling the flow function will return a tensorflow
+    DataSet that contains the GraphWave embeddings.
+    """
+
+    def __init__(self, G, scales="auto", num_scales=3, num_eigenvecs=-1):
+        """
+        Args:
+            G (StellarGraph): the StellarGraph object.
+            scales (str or list of floats): the wavelet scales to use. "auto" will cause the scale values to be
+                automatically calculated.
+            num_scales (int): the number of scales when scales = "auto".
+            num_eigenvecs (int): the number of eigenvectors to use. When set to `-1` the maximum number of eigenvectors
+                is calculated.
+        """
+
+        if not isinstance(G, StellarGraph):
+            raise TypeError("G must be a StellarGraph object.")
+
+        node_types = list(G.node_types)
+        if len(node_types) > 1:
+            raise TypeError(
+                "{}: node generator requires graph with single node type; "
+                "a graph with multiple node types is passed. Stopping.".format(
+                    type(self).__name__
+                )
+            )
+
+        # Create sparse adjacency matrix:
+        adj = G.to_adjacency_matrix().tocoo()
+
+        # Function to map node IDs to indices for quicker node index lookups
+        self._node_lookup = G._get_index_for_nodes
+
+        degree_mat = diags(np.array(adj.sum(1)).flatten())
+        laplacian = degree_mat - adj
+
+        if num_eigenvecs == -1:
+            num_eigenvecs = laplacian.shape[0] - 2
+
+        # TODO: add in option to compute wavelet transform using Chebysev polynomials
+        eigen_vals, eigen_vecs = eigs(laplacian, k=num_eigenvecs)
+        eigen_vals = np.real(eigen_vals).astype(np.float32)
+        self.eigen_vecs = np.real(eigen_vecs).astype(np.float32)
+
+        if scales == "auto":
+
+            e2 = eigen_vals[eigen_vals > 0].min()
+            eN = eigen_vals.max()
+
+            min_scale = -np.log(0.95) / np.sqrt(eN * e2)
+            max_scale = -np.log(0.85) / np.sqrt(eN * e2)
+
+            scales = np.linspace(min_scale, max_scale, num_scales)
+
+        self.scales = scales
+
+        # the columns of U exp(-scale * eigenvalues) U^T (U = eigenvectors) are used to calculate the node embeddings
+        # (each column corresponds to a node)
+        # to avoid computing a dense NxN matrix when only several eigenvalues are specified
+        # U exp(-scale * eigenvalues) is computed and stored - which is an N x num_eigenvectors matrix
+        # the columns of U exp(-scale * eigenvalues) U^T are then computed on the fly in generator.flow()
+        Ues = [
+            self.eigen_vecs.dot(np.diag(np.exp(-s * eigen_vals)))[np.newaxis, :]
```

Comment:
Even ignoring the discussion about optimising `E = num_eigenvecs`, I think the use of `np` matrix operations here is going to be unnecessarily expensive: it has to create a `E x E` matrix with only `E` entries set, and then do an n&times;E * E&times;E (i.e. `O(n E^2)`) matrix multiplication, but I think this can be done in `O(n E)` time by using broadcasting to multiply the columns by `n.exp(-s * eigen_vals)` directly:

```
>>> eigenvecs = np.array([[1, 2], [3, 4], [5, 6]])
>>> eigenvals = np.array([10, 100])
>>> eigenvecs * eigenvals
array([[ 10, 200],
       [ 30, 400],
       [ 50, 600]])
```

That is, I think this can be

```suggestion
            (self.eigen_vecs * np.exp(-s * eigen_vals))[np.newaxis, :]
```

Also, minor thing, but the `[np.newaxis, :]` was confusing to me because it's implying the array has a single dimension due to only having one `:`, however I was pretty sure the array already has 2 dimensions. Thus, I think it would be clearer as either `[np.newaxis, :, :]` or `[np.newaxis, ...]` to hint that the array does indeed start with two dimensions.

Reply:
that's a nice optimization, thanks I'll chuck it in :)